### DISCUSSION TITLE
Connecting to a Redis cluster in GCP memorystore for redis cluster locally

### DISCUSSION THREAD

**Discussion by tituscarl** [redis/rueidis#617] [Q&A]
Hi, does anyone knows or have an idea how to solve this problem? We have a memorystore for redis cluster in GCP, and I want to connect to it locally, I tried this [guide](https://cloud.google.com/memorystore/docs/redis/connect-redis-instance#connecting_from_a_local_machine_with_port_forwarding) and it works using redis-cli but when using a redis-client in go for example rueidis it says connection refuse with an address. I assume that the address on the error message is one of the nodes of the clsuter, but is there any reason that it does not connect while using redis-cli does? I tried PING command on cli and on code but only cli works.

**Comment by rueian** [comment]
Hi @tituscarl,

The link you provided is "Connecting from a local machine with port forwarding". I believe the method only supports standalone redis. redis cluster requires the client to connect to each redis node directly and simple port forwarding is not enough.

**Reply by tituscarl** [reply]
Thanks @rueian for answering. I tried port forwarding multiple ports, to each of my clusters nodes. Like this
`gcloud compute ssh COMPUTE_VM_NAME --zone=ZONE -- -N -L 6379:DISCOVERY_ENDPOINT_IP:6379 -L 7000:NODE_1_IP:7000 -L 7001:NODE_2_IP:7001` and so on, but still doesn't work. Is there any method to connect to my remote cluster nodes locally?

**Reply by rueian** [reply]
I think you are very close, but port forwarding like this is still not enough because a client needs to connect each node in the way that the discovery endpoint provides. For example, the discovery endpoint may say "Hey, you need to connect the nodes A and B by using the address FQDN_OF_NODE_A:6379 and FQDN_OF_NODE_B:6379", then you will need to somehow map these addresses in your local environment to the ports you forwarded, typically by setting up some custom domain name resolution rules, so that the client can work with the cluster.

**Reply by tituscarl** [reply]
For the mapping, I also did this on the code (using go-redis):
```
	client = goredisv9.NewClusterClient(&goredisv9.ClusterOptions{
		Addrs: []string{"localhost:6379", "localhost:7000", "localhost:7001"},
	})
```

**Reply by rueian** [reply]
Thank you for sharing the config you used with goredis, but did it actually work with the discovery endpoint?

**Reply by tituscarl** [reply]
In local no, but no issues in prod with just the discovery endpoint.

**Reply by tituscarl** [reply]
The config I send doesn't work also, the first address maps to the discovery endpoint and the 2 are for the nodes address that is also mapped to the forwarded port.

**Reply by rueian** [reply]
> In local no, but no issues in prod with just the discovery endpoint.

That was expected. What error did you see when using rueidis with the discovery endpoint in prod?

**Reply by tituscarl** [reply]
The issue is only on local, I think the problem is not in rueidis. But yeah although it works in prod, I just want to access it in local for some testing.

**Reply by rueian** [reply]
I understand. However, a client still needs to follow the discovery endpoint. It should connect each redis instance in the way the discovery endpoint requested. So you need to somehow translate the production addresses to your local ports on the fly. Putting local addresses in the initialization config of a client will not work.
